**<p style = "text-align: center">TRƯỜNG ĐẠI HỌC KHOA HỌC TỰ NHIÊN - ĐẠI HỌC QUỐC GIA TPHCM</p>**

# **<p style = "text-align: center">NHẬP MÔN KHOA HỌC DỮ LIỆU</p>**
## **<p style = "text-align: center">Đồ án Cuối kì</p>**
### <p style = "text-align: center"> <b>Đề tài:</b> Phân tích, khám phá dữ liệu về sách trên web thương mại điện tử (tiki)</p>
**Sinh viên thực hiện**
| Mã số sinh viên | Họ và tên |
| --- | --- |
| 20120040 | Nguyễn Quang Gia Bảo |
| 20120136 | Huỳnh Tuấn Nam |
| 20120158 | Trần Hoàng Anh Phi |

---

# **1. Thông tin về dữ liệu**<br>
Dữ liệu được sử dụng trong bài được lấy từ trang: <a href = "https://tiki.vn/sach-truyen-tieng-viet/c316"> https://tiki.vn/sach-truyen-tieng-viet/c316</a><br>

Vì website <a href = 'https://tiki.vn/'> https://tiki.vn/ </a> có khá là nhiều mặt hàng nên nhóm chỉ chọn ra danh mục <font color=red>sách tiếng Việt</font> để crawl và thực hiện đồ án này.


# **2. Khám phá dữ liệu**<br>

#### Import các package cần thiết

In [50]:
import pandas as pd
import numpy as np


#### Đọc dữ liệu đã thu thập từ file csv 

In [51]:
df = pd.read_csv('full_data/data.csv')
print(df.shape)
display(df.head())

C:\Users\ADMIN\AppData\Local\Temp\ipykernel_21608\906639567.py:1: DtypeWarning: Columns (22,27,29,30,31,32,33,34,35,36,37,38,39,40,41,43,45,46,47,48,49,50,51,52,53,54,56,57,58,59,60,61,62,63,64,65,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('full_data/data.csv')


(96647, 84)


Unnamed: 0         id  master_id            sku  \
0           0  204317934  204317934  9972713357587   
1           1  205279963  205279963  7618382614299   
2           2  193209826  193209826  5001141446127   
3           3  205288789  205288789  9035774965791   
4           4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  ...  kich_thuoc  tai_trong device_brand  \
0          325000             19  ...         NaN        NaN          NaN   
1          325000             21  ...         NaN        NaN          NaN   
2          129000             26  ...         NaN        NaN          NaN   
3          325000             10  ...         NaN        NaN          NaN   
4          325000             10  ...         NaN        NaN          NaN   

  do_nhay  tan_so huong_dan_bao_quan  dac_diem_noi_bat dieu_kien_su_dung  \
0     NaN     NaN                NaN               NaN               NaN   
1     NaN     NaN                NaN               NaN               NaN   
2     NaN     NaN                NaN               NaN               NaN   
3     NaN     NaN                NaN               NaN               NaN   
4     NaN     NaN                NaN               NaN               NaN   

  dia_chi_su_dung bluetooth  
0             NaN       NaN  
1             NaN       NaN  
2             NaN       NaN  
3             NaN       NaN  
4             NaN       NaN  

[5 rows x 84 columns]

In [52]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96647 entries, 0 to 96646
Data columns (total 84 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Unnamed: 0                    96647 non-null  int64  
 1   id                            96647 non-null  int64  
 2   master_id                     96647 non-null  int64  
 3   sku                           96647 non-null  int64  
 4   name                          96647 non-null  object 
 5   short_url                     96647 non-null  object 
 6   book_cover                    66029 non-null  object 
 7   price                         96647 non-null  int64  
 8   original_price                96647 non-null  int64  
 9   discount_rate                 96647 non-null  int64  
 10  rating_average                96647 non-null  float64
 11  review_count                  96647 non-null  int64  
 12  inventory_type                96647 non-null  object 
 13  p

***

#### **Cleaning data**

Drop cột 'Unnamed: 0'

In [53]:
df = df.drop(columns='Unnamed: 0')
df.head()

id  master_id            sku  \
0  204317934  204317934  9972713357587   
1  205279963  205279963  7618382614299   
2  193209826  193209826  5001141446127   
3  205288789  205288789  9035774965791   
4  204649516  204649516  8708257676679   

                                                name  \
0     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
1     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
2  Con Đường Tơ Lụa - Vạn Dặm Xa Từ Pakistan Tới ...   
3     1111 - Nhật Ký Sáu Vạn Dặm Trên Yên Xe Cà Tàng   
4  Sách - 1111 - NHẬT KÝ SÁU VẠN DẶM TRÊN YÊN XE ...   

                                           short_url book_cover   price  \
0  https://tiki.vn/product-p204317934.html?spid=2...        NaN  262000   
1  https://tiki.vn/product-p205279963.html?spid=2...    Bìa mềm  256750   
2  https://tiki.vn/product-p193209826.html?spid=1...    Bìa mềm   95100   
3  https://tiki.vn/product-p205288789.html?spid=2...    Bìa mềm  292500   
4  https://tiki.vn/product-p204649516.html?spid=2...        NaN  292500   

   original_price  discount_rate  rating_average  ...  kich_thuoc tai_trong  \
0          325000             19             5.0  ...         NaN       NaN   
1          325000             21             0.0  ...         NaN       NaN   
2          129000             26             4.8  ...         NaN       NaN   
3          325000             10             0.0  ...         NaN       NaN   
4          325000             10             0.0  ...         NaN       NaN   

  device_brand  do_nhay tan_so  huong_dan_bao_quan dac_diem_noi_bat  \
0          NaN      NaN    NaN                 NaN              NaN   
1          NaN      NaN    NaN                 NaN              NaN   
2          NaN      NaN    NaN                 NaN              NaN   
3          NaN      NaN    NaN                 NaN              NaN   
4          NaN      NaN    NaN                 NaN              NaN   

  dieu_kien_su_dung dia_chi_su_dung bluetooth  
0               NaN             NaN       NaN  
1               NaN             NaN       NaN  
2               NaN             NaN       NaN  
3               NaN             NaN       NaN  
4               NaN             NaN       NaN  

[5 rows x 83 columns]

Xem thử dữ liệu có bị trùng hay không?

In [54]:
df.duplicated().sum()

8294

Vậy là dữ liệu có 8294 dòng trùng nhau, ta sẽ tiến hành xoá những dòng này

In [55]:
df = df.drop_duplicates(ignore_index=True)
df.shape

(88353, 83)

Vì khi thu thập dữ liệu, có những sản phẩm khác loại bị tiki xếp vào chung danh mục nên ta thấy những columns lạ bị thêm vào, ta phải tiến hành xoá những columns này và những sản phẩm <font color =red>lạc loài </font>

In [56]:
df.columns

Index(['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'brand', 'brand_country', 'item_model_number',
       'origin', 'product_weight', 'battery_capacity', 'camera', 'chong_nuoc',
       'loai_day', 'san_pham', 'storage', 'battery_life', 'charge_time',
       'do_chiu_nuoc', 'included_accessories', 'material', 'luu_y',
       'minimum_inbound_policy_days', 'expiry_time', 'shelf_life_days',
       'chat_lieu', 'audio_power_output', 'audio_technology',
       'dieu_khien_bang_giong_noi', 'dieu_khien_tivi', 'digital_tivi_dvb_t2',
       'display_type', 'frequency', 'hdh_giao_dien',
       'image_processing_technology', 'network_hdmi', 'network_inte

In [57]:
df[~df['material'].isna()]

id  master_id            sku  \
13233  144356210  144356210  5525114755464   
14886  207035527  207035527  6816910771280   
15160  176689612  176689612  9588934896842   
25077  207631100  207631100  6801021851103   
49830  196565537  196565537  5928507424911   
63694  202432845  202432845  8183965466103   
76899  175150080  175150080  9992881446659   

                                                    name  \
13233  Túi đeo chạy bộ, túi đeo hông, đeo bụng dành c...   
14886  Áo Hoodie Nỉ Ngoại Nữ Thêu Chữ Loang Màu Hồng ...   
15160  Vỏ gối trắng trơn-vỏ gối cao cấp 100% Cotton-t...   
25077     Áo Khoác Nam Nữ Xanh Rêu Áo Khoác Dù Dáng Rộng   
49830  Rèm Tự Dính-Không Cần Khoan Vô Cùng Tiện Lợi, ...   
63694  BBSTORE'S Đầm Xòe Dự Tiệc Tay Phồng Thời Trang...   
76899  Bộ Bàn Ghế Gỗ Me Tây Nguyên Khối, Bàn Ghế Ăn G...   

                                               short_url book_cover     price  \
13233  https://tiki.vn/product-p144356210.html?spid=1...        NaN    112000   
14886  https://tiki.vn/product-p207035527.html?spid=2...        NaN    109000   
15160  https://tiki.vn/product-p176689612.html?spid=1...        NaN     63750   
25077  https://tiki.vn/product-p207631100.html?spid=2...        NaN    130000   
49830  https://tiki.vn/product-p196565537.html?spid=1...        NaN     52500   
63694  https://tiki.vn/product-p202432845.html?spid=2...        NaN    219000   
76899  https://tiki.vn/product-p175150080.html?spid=1...        NaN  20400000   

       original_price  discount_rate  rating_average  ...  \
13233          139000             19             4.8  ...   
14886          109000              0             0.0  ...   
15160           63750              0             0.0  ...   
25077          130000              0             0.0  ...   
49830          105000             50             0.0  ...   
63694          219000              0             0.0  ...   
76899        20400000              0             0.0  ...   

                 kich_thuoc tai_trong device_brand  do_nhay tan_so  \
13233                   NaN       NaN          NaN      NaN    NaN   
14886                   NaN       NaN          NaN      NaN    NaN   
15160                   NaN       NaN          NaN      NaN    NaN   
25077                   NaN       NaN          NaN      NaN    NaN   
49830           30 x 20 x 5       5kg          NaN      NaN    NaN   
63694                   NaN       NaN          NaN      NaN    NaN   
76899  3,39 x (70 - 75) x 5       NaN          NaN      NaN    NaN   

       huong_dan_bao_quan dac_diem_noi_bat dieu_kien_su_dung dia_chi_su_dung  \
13233                 NaN              NaN               NaN             NaN   
14886                 NaN              NaN               NaN             NaN   
15160                 NaN              NaN               NaN             NaN   
25077                 NaN              NaN               NaN             NaN   
49830                 NaN              NaN               NaN             NaN   
63694                 NaN              NaN               NaN             NaN   
76899                 NaN              NaN               NaN             NaN   

      bluetooth  
13233       NaN  
14886       NaN  
15160       NaN  
25077       NaN  
49830       NaN  
63694       NaN  
76899       NaN  

[7 rows x 83 columns]

Các `features` được sử dụng

In [58]:
features = ['id', 'master_id', 'sku', 'name', 'short_url', 'book_cover', 'price',
       'original_price', 'discount_rate', 'rating_average', 'review_count',
       'inventory_type', 'productset_group_name', 'day_ago_created',
       'categories', 'all_time_quantity_sold', 'authors', 'publisher_vn',
       'publication_date', 'dimensions', 'manufacturer', 'number_of_page',
       'dich_gia', 'edition', 'luu_y']

Một số `features` không hợp lệ với yêu cầu bài toán

In [59]:
unvalid_cols = list(set(df.columns) - set(features))
print(unvalid_cols)

['network_internet', 'bluetooth', 'image_processing_technology', 'weight_without_stand', 'brand_country', 'camera', 'do_nhay', 'audio_technology', 'chong_nuoc', 'weight_with_stand', 'dieu_khien_bang_giong_noi', 'screen_size', 'usb', 'storage', 'expiry_time', 'network_wifi', 'chat_lieu', 'screen_mirroring', 'dieu_kien_su_dung', 'remote_thong_minh', 'huong_dan_bao_quan', 'minimum_inbound_policy_days', 'do_chiu_nuoc', 'included_accessories', 'dac_diem_noi_bat', 'frequency', 'audio_power_output', 'origin', 'dieu_khien_tivi', 'san_pham', 'brand', 'tan_so', 'size_without_stand_table_top', 'display_type', 'capacity', 'size', 'size_with_stand_table_top', 'shelf_life_days', 'charge_time', 'digital_tivi_dvb_t2', 'ung_dung', 'tai_trong', 'battery_life', 'thoi_gian_su_dung', 'dia_chi_su_dung', 'material', 'battery_capacity', 'kich_thuoc', 'resolution', 'network_hdmi', 'tivi_type', 'hdh_giao_dien', 'product_weight', 'device_brand', 'loai_pin', 'loai_day', 'huong_dan_su_dung', 'item_model_number']


Thử lấy các sản phẩm thuộc ngoài những `categories` của danh mục sách (mà ta đã thu thập + 3 danh mục mới)

In [60]:
valid_categories = pd.read_csv('id_data/categories_id.csv')['categories'] 
valid_categories = '|'.join(list(valid_categories) + ['Root', 'Sách tiếng Việt', 'Nhà Sách Tiki'])
# df[df[unvalid_cols].notna().any(axis=1)].to_csv('not_books.csv', encoding='utf-8-sig')
# df[df['publisher_vn'].isna()].to_csv('not_publiser.csv', encoding='utf-8-sig')
# df[df['categories'].str.contains(valid_categories, na=False)]

#Chọn những categories không phải sách
df[~df['categories'].str.contains(valid_categories, na=False)]['categories'].value_counts()

Máy Ảnh - Máy Quay Phim                4
Túi và Hộp Đựng Máy Ảnh                3
Thiết Bị Số - Phụ Kiện Số              3
Tai Nghe True Wireless                 2
Thiết Bị Phòng Tối                     1
Thực phẩm bổ dưỡng                     1
Phụ kiện ô tô                          1
Bánh kẹo                               1
Móng giả                               1
Bộ sản phẩm vệ sinh nhà cửa            1
Làm Đẹp - Sức Khỏe                     1
Thực phẩm khô khác                     1
Loa Vi Tính                            1
Dây Đeo Máy Ảnh, Máy Quay, Ống Nhòm    1
Bình, ly uống trà và phụ kiện          1
Nước Ngọt                              1
Bàn ghế làm việc                       1
Nón bơi                                1
Phụ Kiện Máy Quay Khác                 1
Bao Da - Ốp Lưng iPad                  1
Bình đun siêu tốc                      1
Các loại gia vị dạng nước khác         1
Voucher - Dịch vụ                      1
Mystery, Thriller & Suspense           1
Thời Trang      

Và ta chỉ lấy những giá trị hợp lệ 

In [61]:
df = df[df['categories'].str.contains(valid_categories, na=False)]
# df.shape
df.shape

(87955, 83)

Có thể vẫn còn, ta sẽ xem những features bị <font color=red>lạc loài</font> (khác giá trị nan):

In [66]:
unvalid_products = df[df[unvalid_cols].notna().any(axis=1)]
# display(unvalid_products)

Tiếp tục bỏ đi những sản phẩm này.

In [63]:
df = df.drop(unvalid_products.index)
df[df[unvalid_cols].notna().any(axis=1)] # test lại

Empty DataFrame
Columns: [id, master_id, sku, name, short_url, book_cover, price, original_price, discount_rate, rating_average, review_count, inventory_type, productset_group_name, day_ago_created, categories, all_time_quantity_sold, authors, publisher_vn, publication_date, dimensions, manufacturer, number_of_page, dich_gia, edition, brand, brand_country, item_model_number, origin, product_weight, battery_capacity, camera, chong_nuoc, loai_day, san_pham, storage, battery_life, charge_time, do_chiu_nuoc, included_accessories, material, luu_y, minimum_inbound_policy_days, expiry_time, shelf_life_days, chat_lieu, audio_power_output, audio_technology, dieu_khien_bang_giong_noi, dieu_khien_tivi, digital_tivi_dvb_t2, display_type, frequency, hdh_giao_dien, image_processing_technology, network_hdmi, network_internet, network_wifi, remote_thong_minh, resolution, screen_mirroring, screen_size, size_without_stand_table_top, size_with_stand_table_top, tivi_type, ung_dung, usb, weight_without_stand, weight_with_stand, size, loai_pin, thoi_gian_su_dung, capacity, huong_dan_su_dung, kich_thuoc, tai_trong, device_brand, do_nhay, tan_so, huong_dan_bao_quan, dac_diem_noi_bat, dieu_kien_su_dung, dia_chi_su_dung, bluetooth]
Index: []

[0 rows x 83 columns]

Có thể vẫn còn những sản phẩm <font color=red>lạc loài</font> khác, tuy nhiên khó có thể nhận biết và loại bỏ được.

Xóa đi các cột <font color =red> lạc loài </font> đi thôi!

In [71]:
df = df.drop(unvalid_cols, axis = 1)
df.shape

(87935, 25)

---

#### **Exploratory Data Analysis**